### ISSUE TITLE
Incorrect/unexpected permissions are returned from `ls -l` spec if the file is a symlink

### ISSUE THREAD

**Issue by skontar** [RedHatInsights/insights-core#3323]
There are multiple specs which use `ls -l` command to read file permissions and parse them using `FilePermissionsParser`. When the file is moved and symlink is used to point to the new location, permissions of the symlink (i.e. ` lrwxrwxrwx.`) are returned instead of the target file permissions.

Impacted specs seem to be:
```
    grub_config_perms = simple_command("/bin/ls -l /boot/grub2/grub.cfg")  # only RHEL7 and updwards
    ...
    grub1_config_perms = simple_command("/bin/ls -l /boot/grub/grub.conf")  # RHEL6
    ...
    sshd_config_perms = simple_command("/bin/ls -l /etc/ssh/sshd_config")
```

My testing suggest that using a `-H` commandline option fixes the problem and works for both cases – with and without symlink.

System with symlink:
```
[root@skontar grub2]# ls -l /boot/grub2/grub.cfg
lrwxrwxrwx. 1 root root 18 Jan 18 07:06 /boot/grub2/grub.cfg -> ../loader/grub.cfg
[root@skontar grub2]# ls -lH /boot/grub2/grub.cfg
-rw-r--r--. 1 root root 4880 May  7  2019 /boot/grub2/grub.cfg
[root@skontar grub2]# ls -l ../loader/grub.cfg
-rw-r--r--. 1 root root 4880 May  7  2019 ../loader/grub.cfg
```

System without symlink:
```
[root@skontar vagrant]# ls -l /boot/grub2/grub.cfg 
-rw-r--r--. 1 root root 4880 May  7  2019 /boot/grub2/grub.cfg
[root@skontar vagrant]# ls -lH /boot/grub2/grub.cfg 
-rw-r--r--. 1 root root 4880 May  7  2019 /boot/grub2/grub.cfg
```

**Comment by wushiqinlou** [comment]
@skontar - These 3 specs are all used to check permissions in insights-prodsec-rules. After updating these specs with "-H", the resolution from insights-content will let customer to update permission of symbolic file(like "/boot/grub2/grub.cfg", not "../loader/grub.cfg"). However, the result is same. If it is OK to you, I will update these specs with "-H" option. @xiangce - Please let me know your opinion.

**Comment by xiangce** [comment]
@wushiqinlou - sounds good to me.

**Comment by skontar** [comment]
I believe that would work as expected? User is able to treat "/boot/grub2/grub.cfg" as a path to the real file and does not need to know that there is a symlink, am I correct?

See:
```
[root@skontar grub2]# ls -l /boot/grub2/grub.cfg 
lrwxrwxrwx. 1 root root 18 Jan 19 04:22 /boot/grub2/grub.cfg -> ../loader/grub.cfg
[root@skontar grub2]# ls -lH /boot/grub2/grub.cfg 
-rw-r--rw-. 1 root root 4880 May  7  2019 /boot/grub2/grub.cfg
[root@skontar grub2]# chmod o-w /boot/grub2/grub.cfg
[root@skontar grub2]# ls -lH /boot/grub2/grub.cfg 
-rw-r--r--. 1 root root 4880 May  7  2019 /boot/grub2/grub.cfg
```

**Comment by wushiqinlou** [comment]
@skontar - Yes, you are correct. I just want to confirm if you are OK to show the  "/boot/grub2/grub.cfg" in the content instead of real file(../loader/grub.cfg). If you think current content works as expected, we only need to update the specs.

**Comment by jsvob** [comment]
Short answer: My opinion that the change from `/bin/ls -l ` to `/bin/ls -lH ` **is correct**.

Long answer:

There's some context that I think is necessary to be documented for posterity.

The _purpose_ of the specs mentioned by the OP is to read the permissions and ownership of the grub configuration _**files**_. The purpose is _has never been_ reading the whole _chain_ of parent directories as well (wholly separate and different specs would have to be used for that). The reason for that is that the purpose of the specs' use has always been ensuring the _file_ permissions are correct, and never checking the permissions of the parent directories (that would be a different issue, different attack scenario, different threat model).

Therefore, the specs mentioned by the OP should stick to returning permission and owner information of the target _**file**_ even if the original path is only a symbolic link.

The reason why the permissions and ownership of the symbolic links themselves is irrelevant is that in Linux, permissions and ownership of symbolic links are mostly ignored. Changes of symlink contents (targets) are, AFAIK, possible only via `unlink` and `symlink` system calls and never modifiable in-place, and therefore subject to parent directory permissions which are out of scope as mentined above. References: 

* https://superuser.com/a/303063
* man 2 symlink
* man 2 open
* man 2 chown
* man 2 chmod
* man 1 chown
* man 1 chmod

To check that `ls -lH` behaves as expected w.r.t the _purpose_ of the specs mentioned by the OP, here follows a bash script that can be used to check that our assumptions are correct.

```
#!/bin/bash
sudo rm -rf /tmp/linktests
mkdir -p /tmp/linktests/orig/path/to
mkdir -p /tmp/linktests/link/to/orig
mkdir -p /tmp/linktests/link/to/link
echo test > /tmp/linktests/orig/path/to/file
ln -s '/tmp/linktests/orig/path/to/file' /tmp/linktests/link/to/orig/file
ln -s '../../../link/to/orig/file' /tmp/linktests/link/to/link/file
if [[ "$( ls -lan /tmp/linktests/link/to/link/file )" =~ 'lrwxrwxrwx. 1 '.*' /tmp/linktests/link/to/link/file -> ../../../link/to/orig/file' ]] ; then
    echo "test1: OK"
else
    echo "test1: ERR"
fi
if [[ "$( ls -lan /tmp/linktests/link/to/orig/file )" =~ 'lrwxrwxrwx. 1 '.*' /tmp/linktests/link/to/orig/file -> /tmp/linktests/orig/path/to/file' ]] ; then
    echo "test2: OK"
else
    echo "test2: ERR"
fi
if [[ "$( ls -lan /tmp/linktests/orig/path/to/file )" =~ '-rw-rw-r--. 1 '.*' /tmp/linktests/orig/path/to/file' ]] ; then
    echo "test3: OK"
else
    echo "test3: ERR"
fi
if [[ "$( ls -lanH /tmp/linktests/link/to/link/file)" =~ '-rw-rw-r--. 1 '.*' /tmp/linktests/link/to/link/file' ]] ; then
    echo "test4: OK"
else
    echo "test4: ERR"
fi
if [[ "$( ls -lanH /tmp/linktests/link/to/orig/file)" =~ '-rw-rw-r--. 1 '.*' /tmp/linktests/link/to/orig/file' ]] ; then
    echo "test5: OK"
else
    echo "test5: ERR"
fi
if [[ "$( ls -lanH /tmp/linktests/orig/path/to/file )" =~ '-rw-rw-r--. 1 '.*' /tmp/linktests/orig/path/to/file' ]] ; then
    echo "test6: OK"
else
    echo "test6: ERR"
fi

# -H omitted on purpose
old_linkls="$( ls -lan /tmp/linktests/link/to/link/file )"
old_filels="$( ls -lan /tmp/linktests/orig/path/to/file )"
# --no-dereference omitted on purpose
sudo chown root:root /tmp/linktests/link/to/link/file
# the change propagates to the file; links don't have their own permissions
new_linkls="$( ls -lan /tmp/linktests/link/to/link/file )"
new_filels="$( ls -lan /tmp/linktests/orig/path/to/file )"
if [[ "$old_linkls" == "$new_linkls" ]] ; then
    echo "test7: OK"
else
    echo "test7: ERR"
fi
if [[ "$old_filels" != "$new_filels" ]] ; then
    echo "test8: OK"
else
    echo "test8: ERR"
fi
```

**Comment by skontar** [comment]
@wushiqinlou seems that we are in an agreement. Please go forward. Thanks!

**Comment by xiangce** [comment]
Resolved by #3324
